pkg/xds/server/callbacks/dataplane_callbacks.go

```diff
@@ -106,25 +107,26 @@ func (d *xdsCallbacks) OnStreamRequest(streamID core_xds.StreamID, request util_
 	// wasn't processed
 	alreadyProcessed = d.dpStreams[streamID].dp != nil
 	if alreadyProcessed {
+		d.Unlock()
 		return nil
 	}
 
-	dpStream := d.dpStreams[streamID]
-	dpStream.dp = &dpKey
-	d.dpStreams[streamID] = dpStream
-
 	activeStreams := d.activeStreams[dpKey]
-	d.activeStreams[dpKey]++
-	d.Unlock()
-
 	if activeStreams == 0 {
+		dpStream := d.dpStreams[streamID]
+		dpStream.dp = &dpKey
+		d.dpStreams[streamID] = dpStream
+
+		d.activeStreams[dpKey]++
+		d.Unlock()
```

Comment:
I'd do:

```
if activeStream == 0 {
   // dpStreamPlay
}
d.Unlock()
if activeStream == 0 {
		if err := d.callbacks.OnProxyConnected(streamID, dpKey, dpStream.ctx, *metadata); err != nil {
			return err
		}
}
```

Rather than having the `Unlock` in two separate code branch